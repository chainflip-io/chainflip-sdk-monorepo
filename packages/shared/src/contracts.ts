import { Signer, Overrides, ContractTransactionReceipt } from 'ethers';
import { ERC20, ERC20__factory } from './abis';
import { ADDRESSES } from './consts';
import { type ChainflipNetwork, type Asset, Assets } from './enums';

export type TransactionOptions = {
  gasLimit?: bigint;
  gasPrice?: bigint;
  maxFeePerGas?: bigint;
  maxPriorityFeePerGas?: bigint;
  nonce?: number;
  wait?: number;
};

export const extractOverrides = (
  transactionOverrides: TransactionOptions,
): Overrides => {
  const { wait, ...ethersOverrides } = transactionOverrides;

  return ethersOverrides;
};

export const getTokenContractAddress = (
  asset: Asset,
  network: ChainflipNetwork,
): string => {
  if (asset === Assets.FLIP) return ADDRESSES[network].FLIP_CONTRACT_ADDRESS;
  if (asset === Assets.USDC) return ADDRESSES[network].USDC_CONTRACT_ADDRESS;

  throw new Error('Only FLIP and USDC are supported for now');
};

export const getStateChainGatewayContractAddress = (
  network: ChainflipNetwork,
): string => ADDRESSES[network].STATE_CHAIN_GATEWAY_ADDRESS;

export const checkAllowance = async (
  amount: bigint,
  spenderAddress: string,
  erc20Address: string,
  signer: Signer,
) => {
  const erc20 = ERC20__factory.connect(erc20Address, signer);
  const signerAddress = await signer.getAddress();
  const allowance = await erc20.allowance(signerAddress, spenderAddress);
  return { allowance, isAllowable: allowance >= amount, erc20 };
};

export const approve = async (
  amount: bigint,
  spenderAddress: string,
  erc20: ERC20,
  allowance: bigint,
  txOpts: TransactionOptions,
): Promise<ContractTransactionReceipt | null> => {
  if (allowance >= amount) return null;
  const transaction = await erc20.approve(
    spenderAddress,
    amount - allowance,
    extractOverrides(txOpts),
  );
  return transaction.wait(txOpts.wait);
};

export const getVaultManagerContractAddress = (
  network: ChainflipNetwork,
): string => ADDRESSES[network].VAULT_CONTRACT_ADDRESS;

export const getFlipBalance = async (
  network: ChainflipNetwork,
  signer: Signer,
): Promise<bigint> => {
  const flipAddress = getTokenContractAddress('FLIP', network);
  const flip = ERC20__factory.connect(flipAddress, signer);
  return flip.balanceOf(await signer.getAddress());
};
