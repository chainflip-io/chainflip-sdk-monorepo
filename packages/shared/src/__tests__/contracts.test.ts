/* eslint-disable max-classes-per-file */
import { ContractTransaction, VoidSigner } from 'ethers';
import { describe, it, expect, vi } from 'vitest';
import { ERC20__factory } from '../abis/index.js';
import { approve, checkAllowance } from '../contracts.js';

class MockERC20 {
  async allowance(_owner: string, _spender: string): Promise<bigint> {
    throw new Error('unmocked call');
  }

  async approve(_spender: string, _amount: bigint): Promise<ContractTransaction> {
    throw new Error('unmocked call');
  }
}

vi.mock('@/shared/abis/factories/ERC20__factory', () => ({
  ERC20__factory: class {
    static connect() {
      return new MockERC20();
    }
  },
}));
const spender = '0xdeadbeef';

describe(checkAllowance, () => {
  it.each([
    { allowance: 1000n, spend: 100n, expected: true },
    { allowance: 1000n, spend: 1500n, expected: false },
  ])('returns the allowance', async ({ allowance, spend, expected }) => {
    const allowanceSpy = vi
      .spyOn(MockERC20.prototype, 'allowance')
      .mockResolvedValueOnce(BigInt(allowance));
    const signer = new VoidSigner('0xcafebabe');

    const result = await checkAllowance(spend, spender, '0x0', signer);

    expect(result.hasSufficientAllowance).toBe(expected);
    expect(result.allowance).toEqual(BigInt(allowance));
    expect(allowanceSpy.mock.calls).toMatchSnapshot();
  });
});

describe(approve, () => {
  it.each([
    { allowance: 100n, spend: 1000n },
    { allowance: 0n, spend: 1000n },
  ])(
    'approves the spender for an allowance equal to the spend request',
    async ({ allowance, spend }) => {
      const approveSpy = vi.spyOn(MockERC20.prototype, 'approve').mockResolvedValueOnce({
        hash: '0x522acf618f67b097672cbcd5f1d0051cf352b7b4dfec4d51b647ce81b33461e4',
        wait: vi.fn().mockResolvedValue({ status: 1 }),
      } as unknown as ContractTransaction);

      const tx = await approve(
        spend,
        spender,
        ERC20__factory.connect('0x0', new VoidSigner('0xcafebabe')),
        allowance,
        { nonce: 1 },
      );

      expect(tx).toMatchObject({
        hash: '0x522acf618f67b097672cbcd5f1d0051cf352b7b4dfec4d51b647ce81b33461e4',
      });
      expect(approveSpy.mock.calls).toMatchSnapshot();
    },
  );

  it('returns null if the allowance is already sufficient', async () => {
    const tx = await approve(
      10n,
      spender,
      ERC20__factory.connect('0x0', new VoidSigner('0xcafebabe')),
      1000n,
      { nonce: 1 },
    );

    expect(tx).toBe(null);
  });
});
