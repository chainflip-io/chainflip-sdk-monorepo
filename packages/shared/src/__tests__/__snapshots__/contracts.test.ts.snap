// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`approve > approves the spender for an allowance equal to the spend request 1`] = `
[
  [
    "0xdeadbeef",
    900n,
    {
      "nonce": 1,
    },
  ],
]
`;

exports[`approve > approves the spender for an allowance equal to the spend request 2`] = `
[
  [
    "0xdeadbeef",
    1000n,
    {
      "nonce": 1,
    },
  ],
]
`;
